fuel/datasets/hdf5.py

```diff
@@ -64,3 +69,125 @@ def get_data(self, state=None, request=None):
+    driver : str, optional
+        Low-level driver to use. Defaults to `None`. See h5py
+        documentation for a complete list of available options.
+
+    """
+    ref_counts = defaultdict(int)
+
+    def __init__(self, path, which_set=None, subset=None, load_in_memory=False,
+                 driver=None, **kwargs):
+        self.path = path
+        self.which_set = which_set
+        self.subset = subset if subset else slice(None, None, None)
+        self.load_in_memory = load_in_memory
+        self.driver = driver
+        self.load()
+        super(H5PYDataset, self).__init__(**kwargs)
+
+    def _get_file_id(self):
+        file_id = [f for f in self.ref_counts.keys() if f.name == self.path]
+        if not file_id:
+            return self.path
+        file_id, = file_id
+        return file_id
+
+    def load(self):
+        handle = self._out_of_memory_open()
+        self.provides_sources = list(handle.keys())
+        shapes = [data_source.shape for data_source in handle.values()]
+        if any(s[0] != shapes[0][0] for s in shapes):
+            raise ValueError("sources have different lengths")
+        if self.subset.step not in (1, None):
+            raise ValueError("subset.step must be either 1 or None")
+        start, stop = (handle.attrs[self.which_set] if self.which_set
+                       else (0, shapes[0][0]))
+        self.subset = slice(
+            start if self.subset.start is None else self.subset.start,
+            stop if self.subset.stop is None else self.subset.stop,
+            self.subset.step)
+        self.num_examples = self.subset.stop - self.subset.start
+        self.data_sources = ([data_source[self.subset] for
+                              data_source in handle.values()]
+                             if self.load_in_memory else None)
+        self._out_of_memory_close(handle)
+
+    def open(self):
+        return None if self.load_in_memory else self._out_of_memory_open()
+
+    def _out_of_memory_open(self):
+        file_id = self._get_file_id()
+        state = h5py.File(name=file_id, mode="r", driver=self.driver)
+        self.ref_counts[state.id] += 1
+        return state
+
+    def close(self, state):
+        if not self.load_in_memory:
+            self._out_of_memory_close(state)
+
+    def _out_of_memory_close(self, state):
+        self.ref_counts[state.id] -= 1
+        if not self.ref_counts[state.id]:
+            del self.ref_counts[state.id]
+            state.close()
+
+    def get_data(self, state=None, request=None):
+        if self.load_in_memory:
+            return self._in_memory_get_data(state=state, request=request)
+        else:
+            return self._out_of_memory_get_data(state=state, request=request)
+
+    def _in_memory_get_data(self, state=None, request=None):
+        if state is not None or request is None:
+            raise ValueError
+        return self.filter_sources([data_source[request] for data_source
```

Comment:
We can potentially leave this for a later PR, but we should get rid of `filter_sources` at one point. Instead, the `__init__` method should read the `sources` keyword argument and use it to determine which datasets to load from the HDF5 file. `filter_sources` is really a stopgap solution because it assumes all the data/sources are loaded, and then discards the unwanted sources just before returning. This can be very wasteful for large datasets.